contracts/l2/L2DAIWormholeBridge.sol

```diff
@@ -0,0 +1,124 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.7.6;
+pragma abicoder v2;
+
+import {iOVM_L1ERC20Bridge} from "@eth-optimism/contracts/iOVM/bridge/tokens/iOVM_L1ERC20Bridge.sol";
+import {iOVM_L2ERC20Bridge} from "@eth-optimism/contracts/iOVM/bridge/tokens/iOVM_L2ERC20Bridge.sol";
+import {OVM_CrossDomainEnabled} from "@eth-optimism/contracts/libraries/bridge/OVM_CrossDomainEnabled.sol";
+import {OVM_L2CrossDomainMessenger} from "@eth-optimism/contracts/OVM/bridge/messaging/OVM_L2CrossDomainMessenger.sol";
+import {WormholeGUID, WormholeLib} from "../common/LibWormholeGUID.sol";
+import {L1DAIWormholeBridge} from "../l1/L1DAIWormholeBridge.sol";
+
+interface Mintable {
+  function mint(address usr, uint256 wad) external;
+
+  function burn(address usr, uint256 wad) external;
+}
+
+contract L2DAIWormholeBridge is OVM_CrossDomainEnabled {
+  using WormholeLib for WormholeGUID;
+
+  // --- Auth ---
+  mapping(address => uint256) public wards;
+
+  function rely(address usr) external auth {
+    wards[usr] = 1;
+    emit Rely(usr);
+  }
+
+  function deny(address usr) external auth {
+    wards[usr] = 0;
+    emit Deny(usr);
+  }
+
+  modifier auth() {
+    require(wards[msg.sender] == 1, "L2DAITokenBridge/not-authorized");
+    _;
+  }
+
+  event Rely(address indexed usr);
+  event Deny(address indexed usr);
+
+  address public immutable l1Token;
+  address public immutable l2Token;
+  address public immutable l1DAITokenBridge;
+  bytes32 public immutable sourceDomain;
+  mapping(bytes32 => uint256) public batchedDaiToFlush;
+
+  event WormholeInitialized(WormholeGUID wormhole);
+  event Flushed(bytes32 targetDomain, uint256 dai);
+
+  constructor(
+    address _l2CrossDomainMessenger,
+    address _l2Token,
+    address _l1Token,
+    address _l1DAITokenBridge,
+    bytes32 _sourceDomain
+  ) public OVM_CrossDomainEnabled(_l2CrossDomainMessenger) {
+    wards[msg.sender] = 1;
+    emit Rely(msg.sender);
+
+    l2Token = _l2Token;
+    l1Token = _l1Token;
+    l1DAITokenBridge = _l1DAITokenBridge;
+    sourceDomain = _sourceDomain;
+  }
+
+  function initiateWormhole(
+    bytes32 targetDomain,
+    address receiver,
+    uint128 amount,
+    address operator
+  ) external {
+    WormholeGUID memory wormhole = WormholeGUID({
+      sourceDomain: sourceDomain,
+      targetDomain: targetDomain,
+      receiver: receiver,
+      operator: operator,
+      amount: amount,
+      nonce: uint64(OVM_L2CrossDomainMessenger(address(getCrossDomainMessenger())).messageNonce()), // gas optimization, we don't need to maintain our own nonce
```

Comment:
This gas-optimization is to reduce an SLOAD I'm assuming? Does their nonce fit into a 64-bit width uint?

Reply:
It saves the SSTORE cost of incrementing a local storage variable. 

Their nonce is a `uint256`, which could in principle overflow our `uint64` slot after a while but it currently costs about $3 to call the `sendMessage()` L2 method that increments the `messageNonce`, so an attacker would need to spend at least $10^19 to make the nonce overflow our slot. In normal operation, an L2 call to `sendMessage()` is coupled to an ~800,000 gas `relayMessage()` call on L1, which leaves us at least 10^11 years to deploy a new bridge to prevent an overflow :)

Reply:
Ah yes right the increment is saved. This seems probably fine to me. I'm also suggesting we increase the nonce width which will help: https://github.com/makerdao/dss-wormhole/pull/11

I feel like the guys planning IPv4 thinking 32 bits was enough width for all internet addresses :)